scip.proto

```diff
@@ -68,9 +68,13 @@ message ToolInfo {
 
 // Document defines the metadata about a source file on disk.
 message Document {
-  // (Required) Path to the text document relative to the directory supplied in
-  // the associated `Metadata.project_root`. Not URI-encoded. This value should
-  // not begin with a directory separator.
+  // (Required) Unique path to the text document relative to the directory
+  // supplied in the associated `Metadata.project_root`. Not URI-encoded.
+  // This value should not begin with a directory separator.
+  //
+  // TODO: Specify if we allow both '/' and '\' as the path separator
+  // or if we expect Unix-style paths.
+  // TODO: Specify what should be done in the presence of symlinks.
```

Comment:
In SemanticDB we use URI-encoding and maybe that's the best solution. It's simple to explain if we say that joining `Metadata.project_root` and `Document.relative_path` creates a URI to the file path. We should remove the "Not URI-encoded" comment above in that case.

Re symlinks: it's up to the indexer to decide. The consumer can figure out if it's a symlink or not so it's not a problem if `relative_path` points to a symlink.

Reply:
I'm not sure what you mean here in terms of URI encoding: I thought you can't have relative paths in URIs. Would the relative path look like `file:///dir/file`? Or do you mean that it would still be `dir/file`, which is then added as a suffix to the `Metadata.project_root`?

Wrt symlinks: the problem is one of merging `Document` structures. Let's say an `Index` contains a document with path `p` with some occurrences `po` and symbols `ps`. It also contains another document with path `q` with some occurrences `qo` and `qs`. Now if `q` is a symlink to `p`, what will Sourcegraph show when you navigate to `p` via `q`? The results from `qo` and `qs`? The results of merging `po`+`ps` and `qo`+`qs`? Just from `po`+`ps`? Should we flag anything here as a validation error/warning when uploading/ingesting the index?

Reply:
> Or do you mean that it would still be dir/file, which is then added as a suffix to the Metadata.project_root?

This one. We could require that the concatenation of `Metadata.project_root` and `Document.relative_path` makes a valid URI. This is what SemanticDB does but it's been problematic in my experience because very few (Java) libraries make it easy to get out the URI-encoded path part of a URI. In >>>99.9% of the time, it's not a problem because the URI encoding is the same as the non-URI encoding. In practice, I frequently didn't bother with URI-encoding and joined instead the individual path parts with a `/` separator.

Reply:
I'd try something like this.

```proto
(Required) Relative path to the regular file on disk backing this text document.
Must point to a regular file, not a symbolic link.
Must be relative to the associated `Metadata.project_root`.
Must use forward slash separators "/" to join path parts, including on Windows.
Must be canonical, no parts of the path can be ".." and duplicate directory separators "//" are not allowed.
Must not begin with a forward slash "/".
Individual parts of the file path must not be URI-encoded.
```

The original version of SCIP had `Document.uri`, which I think is a lot simpler because you avoid having to document all of these requirements. We turned it into a relative path for the hypothetical use-case of simplifying incremental indexing, but I'm still not convinced this change was necessary for that use-case. The thinking was that the byte payload of a `Document` message should be consistent across machines if we remove absolute paths. 

The problem is that there's no standard for relative URI paths 🤷

Reply:
Updated docs with requirements for paths.